### DISCUSSION TITLE
Clarity needed for securing cache communication for kubernetes stack

### DISCUSSION THREAD

**Discussion by sujaykulkarn** [keycloak/keycloak#28796] [Q&A] [answered]
Hi community,
I am trying to secure cache communication by following this - https://www.keycloak.org/server/caching#_securing_cache_communication 
My Configuration - 
Keycloak - 24.0.2
ENV's set - 
KC_CACHE_STACK = kubernetes
KC_CACHE_EMBEDDED_MTLS_ENABLED = true
KC_CACHE_EMBEDDED_MTLS_KEY_STORE_FILE,KC_CACHE_EMBEDDED_MTLS_KEY_STORE_PASSWORD, KC_CACHE_EMBEDDED_MTLS_TRUST_STORE_FILE, KC_CACHE_EMBEDDED_MTLS_TRUST_STORE_PASSWORD

I have 2 replicas of keycloak in my Kubernetes cluster. 
I am facing the below warning logs in keycloak pods. 

```
2024-04-16 03:22:03,073 WARN  [org.jgroups.protocols.TCP] (TcpServer.Acceptor[7800]-1,testckey-ckey-0-20181) JGRP000006: 192.168.13.163:7800: failed accepting connection from peer SSLSocket[hostname=192.168.95.227, port=37191, Session(1713237723028|TLS_AES_256_GCM_SHA384)]: java.net.SocketException: Socket is closed
2024-04-16 03:22:20,060 WARN  [org.jgroups.protocols.TCP] (TcpServer.Acceptor[7800]-1,testckey-ckey-0-20181) JGRP000006: 192.168.13.163:7800: failed accepting connection from peer SSLSocket[hostname=192.168.95.227, port=48383, Session(1713237740015|TLS_AES_256_GCM_SHA384)]: java.net.SocketException: Socket is closed
```
May I know why these warning logs are coming and also have a query whether TLS needs to be enabled for TCP protocol of Kubernetes Stack? 
@pruivo I saw you have implemented this feature, Can you please help here?
Thanks.

**Comment by pruivo** [comment]
@sujaykulkarn Is there anything else after `Socket is closed` or any exception in pod `192.168.95.227`?  Have you double checked the keystores?

**Reply by sujaykulkarn** [reply]
@pruivo, yeah my issue was with the trust-store part, I was giving an emptyTruststore which was causing the connection to get closed. 

Have a few queries though.
1. After enabling secure cache communication, does it mean the Default port of Jgroups TCP stack is 7800, will it become TLS enabled? I tried with testssl tool against this port, but it says not secure. 
2. Does this enhancement provide mTLS encryption between the members? http://jgroups.org/manual5/index.html#ENCRYPT 
3. Is there a way to provide only TLS encryption, not mTLS? I saw if we don't give Trustore it fails to come up.

**Reply by pruivo** [reply]
Yes, those options should enable data encryption between Keycloak servers. What tool are you using? 

Currently, we configure the TLS as follows: https://github.com/keycloak/keycloak/blob/c0b493406d0223e37f63b0e6c8741967e39f8f6a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java#L268-L277

Those options are the ones documented in JGroups: http://jgroups.org/manual5/index.html#TLS
Keycloak does not allow customization via CLI options, but nothing prevents you from creating a custom Infinispan configuration XML file with different JGroups settings that fit your use case.

**Reply by sujaykulkarn** [reply]
Hi @pruivo, Thanks for the replies. 
We are using this tool to test whether the port is TLS-enabled or not. - https://testssl.sh/testssl.sh, We can run this script against a pod-ip as well.  ex- ./testssl.sh 192.168.2.187:7800 

If I understood correctly, we are enabling TLS for the TCP stack and not explicitly doing encryption right? As it's mentioned in Jgroups Doc - http://jgroups.org/manual5/index.html#ENCRYPT 
I was confused when you said data encryption happens between keycloak servers when the options are enabled, Can you please clarify on this, how data encryption happens?
Sorry if I have very basic queries, New to Infinispan and jgroups part. :) 

For Custom cache till now we are using a custom stack to enable TLS for TCP protocol in Kubernetes Stack. 

> TCP bind_addr="${jgroups.bind.address:GLOBAL}"
>                         bind_port="${jgroups.bind.port,jgroups.tcp.port:7800}"
>                         diag.enabled="${jgroups.diag.enabled:false}"
>                         thread_naming_pattern="pl"
>                         send_buf_size="640k"
>                         sock_conn_timeout="300"
>                         logical_addr_cache_expiration="360000"
>                         port_range="0"
>                         bundler_type="transfer-queue"
>                         bundler.max_size="${jgroups.bundler.max_size:64000}"
>                         thread_pool.min_threads="${jgroups.thread_pool.min_threads:0}"
>                         thread_pool.max_threads="${jgroups.thread_pool.max_threads:200}"
>                         thread_pool.keep_alive_time="60000"
>                         thread_pool.thread_dumps_threshold="1"
>                         tls.enabled="true"
>                         tls.keystore_path="${env.KC_HTTPS_KEY_STORE_FILE}"
>                         tls.keystore_password="${env.KC_HTTPS_KEY_STORE_PASSWORD}"

So we just wanted to check whether the new flags can be used instead of giving our custom cache file, Our use case is we want to enable TLS or mTLS for jgroups.bind.port and the port should be secure.

**Reply by pruivo** [reply]
I tried locally with a Keycloak instance using these options: `./bin/kc.sh start --http-enabled true --hostname-strict false --cache=ispn --cache-embedded-mtls-enabled=true --cache-embedded-mtls-key-store-file=/tmp/certs/site1-keystore.p12 --cache-embedded-mtls-key-store-password=secret --cache-embedded-mtls-trust-store-file=/tmp/certs/ca-keystore.p12 --cache-embedded-mtls-trust-store-password=caSecret -Djgroups.bind_addr=127.0.0.1 --cache-stack=tcp`

The `testssl` tool reported the following:
```
 Testing protocols via sockets except NPN+ALPN 

 SSLv2      not offered (OK)
 SSLv3      not offered (OK)
 TLS 1      not offered
 TLS 1.1    not offered
 TLS 1.2    not offered
 TLS 1.3    offered (OK): final
 NPN/SPDY   not offered
 ALPN/HTTP2 not offered

 Testing cipher categories 

 NULL ciphers (no encryption)                  not offered (OK)
 Anonymous NULL Ciphers (no authentication)    not offered (OK)
 Export ciphers (w/o ADH+NULL)                 not offered (OK)
 LOW: 64 Bit + DES, RC[2,4] (w/o export)       not offered (OK)
 Triple DES Ciphers / IDEA                     not offered
 Obsolete CBC ciphers (AES, ARIA etc.)         not offered
 Strong encryption (AEAD ciphers)              offered (OK)


 Testing robust (perfect) forward secrecy, (P)FS -- omitting Null Authentication/Encryption, 3DES, RC4 

 PFS is offered (OK)          TLS_AES_256_GCM_SHA384 TLS_CHACHA20_POLY1305_SHA256 TLS_AES_128_GCM_SHA256 
 Elliptic curves offered:     prime256v1 secp384r1 secp521r1 X25519 X448 
 Finite field group:          ffdhe2048 ffdhe3072 ffdhe4096 ffdhe6144 ffdhe8192

 Testing server preferences 

 Has server cipher order?     no (TLS 1.3 only)
 Negotiated protocol          TLSv1.3
 Negotiated cipher            TLS_AES_256_GCM_SHA384, 253 bit ECDH (X25519) (limited sense as client will pick)
 Negotiated cipher per proto  (limited sense as client will pick)
     TLS_AES_256_GCM_SHA384:        TLSv1.3
 No further cipher order check has been done as order is determined by the client
```

Aren't you seeing a similar output?

**Reply by pruivo** [reply, marked as answer]
The encryption is performed by the JVM using the certificates to authenticate the connections. It uses SSL sockets, as shown in the error message you pasted: `failed accepting connection from peer SSLSocket`, which encrypts everything sent. 

The only difference between the Keycloak options and your custom configuration is that the CLI options are more strict and reject connections if they do not trust the certificate.

**Reply by sujaykulkarn** [reply]
Hi @pruivo,
Did you pass any extra parameters for the testssl tool, like ca or any ssl-native options?
For me, it gives this "192.168.95.223:7800 doesn't seem to be a TLS/SSL enabled server"

> [root@tc0056node04 cloud-user]# ./testssl.sh 192.168.95.223:7800
> 
> ATTENTION: No cipher mapping file found!
> Please note from 2.9 on testssl.sh needs files in "$TESTSSL_INSTALL_DIR/etc/" to function correctly.
> 
> Type "yes" to ignore this warning and proceed at your own risk --> yes
> 
> ATTENTION: No TLS data file found -- needed for socket-based handshakes
> Please note from 2.9 on testssl.sh needs files in "$TESTSSL_INSTALL_DIR/etc/" to function correctly.
> 
> Type "yes" to ignore this warning and proceed at your own risk --> yes
> 
> No engine or GOST support via engine with your /bin/openssl
> 
> ###########################################################
>     testssl.sh       3.0.8 from https://testssl.sh/
> 
>       This program is free software. Distribution and
>              modification under GPLv2 permitted.
>       USAGE w/o ANY WARRANTY. USE IT AT YOUR OWN RISK!
> 
>        Please file bugs @ https://testssl.sh/bugs/
> 
> ###########################################################
> 
>  Using "OpenSSL 1.0.2k-fips  26 Jan 2017" [~118 ciphers]
>  on tc0056node04:/bin/openssl
>  (built: "reproducible build, date unspecified", platform: "linux-x86_64")
> 
> 
>  Start 2024-04-18 08:37:27        -->> 192.168.95.223:7800 (192.168.95.223) <<--
> 
>  rDNS (192.168.95.223):  testckey-ckey-1.testckey-ckey-headless.ncms.svc.cluster.local. 192-168-95-223.testckey-ckey.ncms.svc.cluster.local.
> 
>  192.168.95.223:7800 doesn't seem to be a TLS/SSL enabled server
>  The results might look ok but they could be nonsense. Really proceed ? ("yes" to continue) --> yes
>  Service detected:       Couldn't determine what's running on port 7800, assuming no HTTP service => skipping all HTTP checks
> 
> 
>  Testing protocols via sockets except NPN+ALPN
> 
>  SSLv2      not offered (OK)
>  SSLv3      not offered (OK)
>  TLS 1      not offered
>  TLS 1.1    not offered
>  TLS 1.2    not offered
>  TLS 1.3    not offered
> 
> You should not proceed as no protocol was detected. If you still really really want to, say "YES" --> yes

The only difference I am seeing is I am using "--cache-stack=kubernetes" but Kubernetes stack in turn uses TCP protocol only. Any idea what I might be missing here?

**Reply by pruivo** [reply]
Nothing is missing as far as I can tell. The `kubernetes` stack works fine locally too.

Can you share the logs for the pod with IP 192.168.95.223 or search for the following message?

`INFO  [org.keycloak.quarkus.runtime.storage.legacy.infinispan.CacheManagerFactory] (ForkJoinPool.commonPool-worker-1) MTLS enabled for communications for embedded caches`

**Reply by sujaykulkarn** [reply]
Hi @pruivo,
Yes, I do see the message in the logs. I was able to check with testssl.sh, I am getting the same response as yours, Thank you. 
For this statement "The only difference between the Keycloak options and your custom configuration is that the CLI options are more strict and reject connections if they do not trust the certificate." For my custom configuration, I haven't specified the tls.client_auth is that the reason for not rejecting the connections?

**Reply by pruivo** [reply]
Yes.

The "server-side" of a TLS/SSL connection can perform validation on the node trying to connect (mutual authentication). There are 3 possible values:
* `none` -> it means accept all and no validation is performed and a malicious Keycloak server can join your cluster.
* `want` -> it asks for validation but if the client does not provide one it accepts, leading to the same problem above. If the client sends its certificate, it is validated against the trust store.
* `need` -> this is forced when using the Keycloak CLI options. The client *must* provide a certificate and it is validated against the trust store. The connection is rejected if the certificate is not provided or valid. A malicious Keycloak server is unable to join your cluster.

Please close the question if it is answered.

**Reply by sujaykulkarn** [reply]
Thankyou @pruivo for clarifying.
